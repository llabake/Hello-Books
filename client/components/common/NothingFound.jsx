import React from 'react';
import PropTypes from 'prop-types';

const NothingFound = ({ text, id }) => {
  return (
    <div
      className="card-panel row center-align"
      id={id}>
      <h4>
        {text}
      </h4>
    </div>
  )
};

NothingFound.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string
};

export default NothingFound;