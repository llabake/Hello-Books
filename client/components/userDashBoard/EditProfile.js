import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import Footer from '../common/Footer';
import { getUser } from '../../helpers/utils';
import TextInput from '../common/TextInput';
import { getUserProfile, editProfile } from '../../actions/userAction';
import InputValidator from '../../helpers/inputValidator';

/**
 * 
 * 
 * @class EditProfile
 * @extends {Component}
 */
class EditProfile extends Component {
  /**
   * Creates an instance of EditProfile.
   * @param {any} props 
   * @memberof EditProfile
   */
  constructor(props) {
    super(props);
    this.state = {
      image: '',
      errors: {},
      isValid: false,
      saving: false,
      uploadedImage: null

    }
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFileChange = this.handleFileChange.bind(this);
  }


  componentWillMount() {
    this.props.getUserProfile().then(() => {
      const { profileToEdit } = this.props;
      this.setState({
        firstName: profileToEdit.firstName,
        lastName: profileToEdit.lastName,
        image: profileToEdit.image
      })
    }).catch(() => { })
  }



  componentDidMount() {

  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    }, () => this.validate())
  }

  handleSubmit(event) {
    event.preventDefault();
    const userData = this.state;
    this.props.editProfile(userData)
  }

  handleFileChange(event) {
    const uploadedImage = event.target.files[0];
    this.setState({
      uploadedImage
    }, () => this.validate())
  }

  validate() {
    const { errors, isValid } = InputValidator.updateProfile(this.state);
    console.log(errors)
    this.setState({ isValid, errors });
    return isValid;
  }

  /**
   * 
   * 
   * @returns {Object} User proile with imgae
   * @memberof EditProfile
   */
  render() {
    const { errors, isValid, saving } = this.state;
    return (
      <div>
        <div id="banner">
          <div className="container form-style signin-form">
            <div className="row">
              <div className="row signup-head">
                <div className="col s12">
                  <h5 className="center-align"> Edit Profile </h5>
                </div>
              </div>
              <div className="col s12 ">
                <div className="row ">
                  <form className="col s12 signup" onSubmit={this.handleSubmit}>
                    <TextInput
                      id='first_name'
                      type='text'
                      icon='account_circle'
                      name='firstName'
                      placeholder='First Name'
                      onChange={this.handleChange}
                      value={this.state.firstName}
                      errors={errors.firstName}
                    />
                    <TextInput
                      id='last_name'
                      type='text'
                      icon='account_circle'
                      name='lastName'
                      placeholder='Last Name'
                      onChange={this.handleChange}
                      clearError={this.clearError}
                      value={this.state.lastName}
                      errors={errors.lastName}
                    />
                    <div className="file-field input-field" style={{ top: '1em' }}>
                      <div className="btn"
                        style={{ width: '40%', fontSize: '13px', marginLeft: '1.8em' }}>
                        <span>Photo</span>
                        <input type="file" accept="image/*"
                          onChange={this.handleFileChange} />
                      </div>
                      <div className="file-path-wrapper">
                        <input className="file-path validate"
                          type="text" style={{ 'width': '89%' }} />
                      </div>
                      {errors.image && errors.image.length ?
                        errors.image.map((error, i) => {
                          return (
                            <div key={i} className='red-text'>
                              <i className="material-icons">error_outline</i>
                              {error}
                            </div>
                          )
                        }) : null}
                    </div>
                    <button
                      type="submit"
                      className="waves-effect waves-light btn"
                      disabled={!isValid || saving}
                      style={{ top: '4em' }}>Save</button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    profileToEdit: state.userReducer.profile,
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    getUserProfile: () => dispatch(getUserProfile()),
    editProfile: (userData) => dispatch(editProfile(userData))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditProfile);
