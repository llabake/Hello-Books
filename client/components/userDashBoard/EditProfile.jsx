import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';

import Footer from '../common/Footer';
import { getUser } from '../../helpers/utils';
import TextInput from '../common/TextInput';
import { getUserProfile, editProfile } from '../../actions/userAction';
import InputValidator from '../../helpers/inputValidator';

/**
 * 
 * 
 * @class EditProfile
 * @extends {Component}
 */
class EditProfile extends Component {
  /**
   * Creates an instance of EditProfile.
   * @param {any} props 
   * @memberof EditProfile
   */
  constructor(props) {
    super(props);
    this.state = {
      image: '',
      errors: {},
      isValid: false,
      saving: false,
      uploadedImage: null,
      redirect: false

    }
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFileChange = this.handleFileChange.bind(this);
  }

  /**
   * @returns {void}
   * 
   * @memberof EditProfile
   */
  componentDidMount() {
    this.props.getUserProfile()
  }

  /**
   * @returns {object} profile to edit
   * 
   * @param {any} nextProps 
   * @memberof EditProfile
   */
  componentWillReceiveProps(nextProps) {
    if(nextProps === this.props) return;
    const { profileToEdit, updated } = nextProps;
    if(updated) {
      setTimeout(() => {
        this.setState({
          redirect: true,
        })
      }, 2000)
    }
    this.setState({
      firstName: profileToEdit.firstName,
      lastName: profileToEdit.lastName,
      image: profileToEdit.image
    })
  }

  /**
   * @returns {object} object containing state
   * 
   * @param {any} event 
   * @memberof EditProfile
   */
  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    }, () => this.validate())
  }
  /**
   * @returns {object} edited profile response
   * 
   * @param {any} event 
   * @memberof EditProfile
   */
  handleSubmit(event) {
    event.preventDefault();
    const userData = this.state;
    this.props.editProfile(userData)
  }
  /**
   * @returns {files} uploaded image
   * 
   * @param {any} event 
   * @memberof EditProfile
   */
  handleFileChange(event) {
    const uploadedImage = event.target.files[0];
    this.setState({
      uploadedImage
    }, () => this.validate())
  }
  /**
   * 
   * 
   * @returns {object} validation response status
   * @memberof EditProfile
   */
  validate() {
    const { errors, isValid } = InputValidator.updateProfile(this.state);
    this.setState({ isValid, errors });
    return isValid;
  }

  /**
   * 
   * 
   * @returns {Object} User proile with imgae
   * @memberof EditProfile
   */
  render() {
    const { errors, isValid, saving, redirect } = this.state;
    return (
      redirect ? 
      <Redirect to='/profile' /> :
      <div>
        <div id="banner">
          <div className="container form-style signin-form">
            <div className="row">
              <div className="row signup-head">
                <div className="col s12">
                  <h5 className="center-align"> Edit Profile </h5>
                </div>
              </div>
              <div className="col s12 ">
                <div className="row ">
                  <form className="col s12 signup" onSubmit={this.handleSubmit}>
                    <TextInput
                      id='first_name'
                      type='text'
                      icon='account_circle'
                      name='firstName'
                      placeholder='First Name'
                      onChange={this.handleChange}
                      value={this.state.firstName}
                      errors={errors.firstName}
                    />
                    <TextInput
                      id='last_name'
                      type='text'
                      icon='account_circle'
                      name='lastName'
                      placeholder='Last Name'
                      onChange={this.handleChange}
                      clearError={this.clearError}
                      value={this.state.lastName}
                      errors={errors.lastName}
                    />
                    <div className="file-field input-field" style={{ top: '1em', marginTop: 110 }}>
                      <div className="btn"
                        style={{ width: '40%', fontSize: '13px', marginLeft: '1.8em' }}>
                        <span>Photo</span>
                        <input type="file" accept="image/*"
                          onChange={this.handleFileChange} />
                      </div>
                      <div className="file-path-wrapper">
                        <input className="file-path validate"
                          type="text" style={{ 'width': '89%', marginTop: 9 }} />
                      </div>
                      {errors.image && errors.image.length ?
                        errors.image.map((error, i) => {
                          return (
                            <div key={i} className='red-text'>
                              {error}
                            </div>
                          )
                        }) : null}
                    </div>
                    <button
                      type="submit"
                      className="waves-effect waves-light btn"
                      disabled={!isValid || saving}
                      style={{ top: '4em' }}>Save</button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    profileToEdit: state.userReducer.profile,
    updated: state.userReducer.updated
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    getUserProfile: () => dispatch(getUserProfile()),
    editProfile: (userData) => dispatch(editProfile(userData))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditProfile);
