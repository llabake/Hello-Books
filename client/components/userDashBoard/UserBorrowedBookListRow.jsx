import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';

import { returnBookAction } from '../../actions/userAction';

/**
 * 
 * 
 * @class UserBorrowedBookListRow
 * @extends {Component}
 */
class UserBorrowedBookListRow extends Component {

  /**
   * Creates an instance of UserBorrowedBookListRow.
   * @param {any} props 
   * @memberof UserBorrowedBookListRow
   */
  constructor(props) {
    super(props);
    this.handleReturn = this.handleReturn.bind(this);
  }

  /**
   * @returns {void}
   * 
   * @param {any} event 
   * @memberof UserBorrowedBookListRow
   */
  handleReturn(event) {
    event.preventDefault();
    this.props.returnBookAction(this.props.borrowedBook.bookId);
  }

  /**
   * @returns {object} borrowedBook object
   * 
   * @memberof UserBorrowedBookListRow
   */
  render() {
    const { borrowedBook, index } = this.props
    return (
      <tr>
        <td>{index + 1}</td>
        <td>
          <Link to={'/book/' + borrowedBook.book.id}>
            {borrowedBook.book.title}
          </Link>
        </td>
        <td>{borrowedBook.book.author}</td>
        <td>{borrowedBook.borrowDate ?
          moment(borrowedBook.borrowDate).format('LL') :
          borrowedBook.borrowDate} </td>
        <td>{borrowedBook.expectedReturnDate ?
          moment(borrowedBook.expectedReturnDate).format('ll') :
          borrowedBook.expectedReturnDate}</td>
        <td>
          {borrowedBook.returnStatus == 'accepted' ? 'Returned' :
            <a
              className={`btn-flat 
              ${borrowedBook.returnStatus === 'pending' ? 'disabled' : null}`}
              onClick={this.handleReturn}>
              <i className="material-icons">
                assignment_returned
            </i>
            </a>
          }
        </td>
      </tr>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    returnBookAction: (bookId) => dispatch(returnBookAction(bookId)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(UserBorrowedBookListRow);
