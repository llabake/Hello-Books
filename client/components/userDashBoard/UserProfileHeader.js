import React, { Component } from 'react';
import { Link, } from 'react-router-dom';
import { Redirect } from 'react-router';
import { connect } from 'react-redux';
import { logout } from '../../actions/userAction';

/**
 * 
 * 
 * @class UserProfileHeader
 * @extends {Component}
 */
class UserProfileHeader extends Component {
  /**
   * Creates an instance of UserProfileHeader.
   * @param {any} props 
   * @memberof UserProfileHeader
   */
  constructor(props) {
    super(props);
    this.state = {
      redirect: false
    }
    this.handleLogout = this.handleLogout.bind(this);
  }

 
  /**
   * 
   * 
   * @returns {object} redirects to home page
   * @memberof UserProfileHeader
   */
  handleLogout() {
    this.props.logout();
    this.setState({
      redirect: true
    })
    this.redirect ? <Redirect to='/' /> : null
  }

  /**
   * 
   * 
   * @returns {object} void
   * @memberof UserProfileHeader
   */
  render () {
    const { user, authenticated } = this.props
    return (
      <div className="navbar-fixed">
        <nav>
          <div className="nav-wrapper">
            <Link to="/" className="brand-logo left adjust">HelloBooks</Link>
            <a href="#" data-activates="mobile-demo" className="button-collapse"><i className="material-icons">menu</i></a>
            
            <ul id="nav-mobile" className="right hide-on-med-and-down">
              <li><Link to="/favorite">Favorite Books</Link></li>
              {/* <!-- <i className="material-icons prefix">notifications</i> --> */}
              <li><Link to="/notifications">Notifications<span className="new badge red">4</span></Link></li>
              <li><a className="dropdown-button" href="#" data-activates="dropdown1">{ authenticated ? user.username : ""}<i className="material-icons right">arrow_drop_down</i></a>
                {/* <!-- Dropdown Structure --> */}
                <ul id="dropdown1" className="dropdown-content">
                  <li><Link to="/editprofile"><i className="material-icons ">settings</i>Profile Setting</Link></li>
                  <li><a href="#!">Terms and Condition</a></li>
                  <li className="divider"></li>
                  <li><Link to=""  onClick={this.handleLogout}><i className="material-icons ">lock</i>Log Out</Link></li>
                </ul>
              </li>
            </ul>
          </div>
        </nav>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    user: state.userReducer.authUser
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    logout: () => dispatch(logout())
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(UserProfileHeader)
