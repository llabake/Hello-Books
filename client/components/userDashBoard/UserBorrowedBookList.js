import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';


import UserBorrowedBookListRow from '../userDashBoard/UserBorrowedBookListRow';
import { fetchUserBorrowedBooks } from '../../actions/userAction';
import { maxPageLimit } from '../../helpers/utils';

/**
 * 
 * 
 * @class UserBorrowedBookList
 * @extends {Component}
 */
class UserBorrowedBookList extends Component {

  /**
   * Creates an instance of UserBorrowedBookList.
   * @param {any} props 
   * @memberof UserBorrowedBookList
   */
  constructor(props) {
    super(props);
    this.state = {
      maxItems: 2,
      showPagination: false,
      activePage: 1,
    }

    this.handleSelectedPage = this.handleSelectedPage.bind(this)

  }

  /**
   * @returns {object} void
   * 
   * @memberof UserBorrowedBookList
   */
  componentDidMount() {
    this.props.fetchUserBorrowedBooks();
  }

  /**
   * 
   * 
   * @param {any} newProps 
   * @returns {Array} response containing user's borrowed book list
   * @memberof UserBorrowedBookList
   */
  componentWillReceiveProps(newProps) {
    if (newProps === this.props) return;
    const { userBorrowedBookCount } = newProps;
    const maxItems = Math.ceil(userBorrowedBookCount / maxPageLimit);
    if (maxItems > 1) {
      this.setState({
        maxItems: maxItems,
        showPagination: true
      })
    }
  }
  /**
   * @returns {void}
   * 
   * @param {any} page 
   * @memberof UserBorrowedBookList
   */
  handleSelectedPage(page) {
    this.props.fetchUserBorrowedBooks(page, maxPageLimit);
  }
  /**
   * @returns {object} void
   * 
   * @memberof UserBorrowedBookList
   */
  render() {
    const { borrowedBooks, userBorrowedBookCount } = this.props;
    const { showPagination } = this.state;
    return (
      <div>
        {borrowedBooks && borrowedBooks.length ?
          <div className="card-panel">
            <table className="bordered centered highlight responsive-table">
              <thead>
                <tr>
                  <th>S/N</th>
                  <th>Title</th>
                  <th>Author </th>
                  <th>Date Borrowed</th>
                  <th>Return By</th>
                  <th>Return</th>
                </tr>
              </thead>
              <tbody>
                {borrowedBooks ? borrowedBooks.map((borrowedBook, i) =>
                  <UserBorrowedBookListRow
                    key={borrowedBook.id}
                    borrowedBook={borrowedBook}
                    index={i}
                  />
                ) :
                  null
                }
              </tbody>
            </table>

            {showPagination ?
              <Pagination
                className={'center-align'}
                items={this.state.maxItems}
                activePage={1} maxButtons={4}
                onSelect={this.handleSelectedPage}
              /> :
              null}
          </div> :
          !userBorrowedBookCount ?
          <div className="card-panel row">
            <p>
              Start borrowing books today!!!
            </p>
          </div> 
          : null
        }
      </div>
    )

  }
}

const mapStateToProps = (state) => {
  return {
    borrowedBooks: state.userReducer.borrowedBookHistory,
    userBorrowedBookCount: state.userReducer.userBorrowedBookCount
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchUserBorrowedBooks: (page) => dispatch(fetchUserBorrowedBooks(page))
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(UserBorrowedBookList);
