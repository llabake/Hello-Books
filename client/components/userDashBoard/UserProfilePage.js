
import React, { Component } from 'react';
import { Link, } from 'react-router-dom';
import { connect } from 'react-redux'

import Footer from '../common/Footer';
import UserProfileSidePanel from '../userDashBoard/UserProfileSidePanel';
import UserBorrowedBookList from '../userDashBoard//UserBorrowedBookList';
import { getUserProfile } from '../../actions/userAction';

/**
 * 
 * 
 * @class UserProfilePage
 * @extends {Component}
 */
class UserProfilePage extends Component {
  /**
   * Creates an instance of UserProfilePage.
   * @param {any} props 
   * @memberof UserProfilePage
   */
  constructor(props) {
    super(props)
  }

  /**
   * @returns {void}
   * 
   * @memberof UserProfilePage
   */
  componentWillMount() {
    this.props.getUserProfile()
  }
  /**
   * 
   * 
   * @returns {object} void
   * @memberof UserProfilePage
   */
  render () {
    const { user } = this.props;
    return (
      <div>
        <div className="row ">
            <UserProfileSidePanel user={user}/>
            <div className="col s12 m8">
              <UserBorrowedBookList />
            </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    user: state.userReducer.authUser
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    getUserProfile: () => dispatch(getUserProfile()),
  }
};

export default connect(mapStateToProps, mapDispatchToProps) (UserProfilePage);
