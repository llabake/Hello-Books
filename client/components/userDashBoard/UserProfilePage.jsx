
import React, { Component } from 'react';
import { Link, } from 'react-router-dom';
import { connect } from 'react-redux'

import Footer from '../common/Footer';
import UserProfileSidePanel from '../userDashBoard/UserProfileSidePanel';
import UserBorrowedBookList from '../userDashBoard//UserBorrowedBookList';
import { getUserProfile } from '../../actions/userAction';

/**
 * 
 * 
 * @class UserProfilePage
 * @extends {Component}
 */
export class UserProfilePage extends Component {
  /**
   * Creates an instance of UserProfilePage.
   * @param {any} props 
   * @memberof UserProfilePage
   */
  constructor(props) {
    super(props)
  }

  /**
   * @returns {void}
   * 
   * @memberof UserProfilePage
   */
  componentDidMount() {
    this.props.getUserProfile()
  }
  /**
   * 
   * 
   * @returns {object} void
   * @memberof UserProfilePage
   */
  render () {
    const { user } = this.props;
    return (
      <div>
        <div className="row ">
            <UserProfileSidePanel user={user}/>
            <div className="col s12 m8">
              <UserBorrowedBookList />
            </div>
        </div>
      </div>
    )
  }
}

export const mapStateToProps = (state) => {
  return {
    user: state.userReducer.authUser
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    getUserProfile: () => dispatch(getUserProfile()),
  }
};

export default connect(mapStateToProps, mapDispatchToProps) (UserProfilePage);
