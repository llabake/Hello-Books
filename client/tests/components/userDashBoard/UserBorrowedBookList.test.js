import React from 'react';

import {
  UserBorrowedBookList,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/userDashBoard/UserBorrowedBookList';

import adminData from '../../mocks/userData';
import userData from '../../mocks/userData';
import NothingFound from '../../../components/common/NothingFound';
import { Pagination } from 'react-materialize';
import { UserBorrowedBookListRow } from '../../../components/userDashBoard/UserBorrowedBookListRow';

function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    handleShowBookList: jest.fn(),
    fetchUserBorrowedBooks: jest.fn(),
    errors: {},
    handleSelectedPage: jest.fn(),
    userBorrowedBookCount: 0,
    borrowedBooks: [],
  };

  const state = {
    maxItems: 2,
    showPagination: false,
    activePage: 1,

  };
  const wrapper = shallow(<UserBorrowedBookList {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('The User Borrowed Book List table', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should call componentDidMount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });
    spy = jest.spyOn(UserBorrowedBookList.prototype, 'componentDidMount')
    const { wrapper, props, state } = setup();
    expect(UserBorrowedBookList.prototype.componentDidMount).toHaveBeenCalledTimes(1)
  })

  test('should render the user borrowed list', () => {
    const { wrapper, props, state } = setup();
    expect(wrapper.find(Pagination).length).toBe(0);
  })

  test('should render nothing found component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      borrowedBooks: [],
      userBorrowedBookCount: 0,
      fetchUserBorrowedBooks: jest.fn(() => Promise.resolve(1)),
    }
    const wrapper = shallow(<UserBorrowedBookList {...props} />);

    expect(wrapper.find(NothingFound).length).toBe(1)

  })

  test('should render UserBorrowedBookList component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      borrowedBooks: userData.borrowedBookList,
      userBorrowedBookCount: 2,
      fetchUserBorrowedBooks: jest.fn(() => Promise.resolve(1)),
    }
    const wrapper = shallow(<UserBorrowedBookList {...props} />);

    expect(wrapper.find(UserBorrowedBookListRow)).toBeDefined();

  })
  test('should call handle selected page', () => {
    const { wrapper, props, state } = setup();
    const page = 2
    wrapper.setState({
      activePage: page,
    });
    wrapper.instance().handleSelectedPage(page)
    expect(wrapper.instance().state.activePage).toBe(2);
  });

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      borrowedBooks: [],
      userBorrowedBookCount: 0,
      fetchUserBorrowedBooks: jest.fn(() => Promise.resolve(1)),
    }

    const nextProps = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      borrowedBooks: [],
      userBorrowedBookCount: 0,
      fetchUserBorrowedBooks: jest.fn(() => Promise.resolve(1)),
    }
    spy = jest.spyOn(UserBorrowedBookList.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<UserBorrowedBookList {...props} />)
    wrapper.setProps(nextProps);
    expect(UserBorrowedBookList.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
  })

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchUserBorrowedBooks');

    const { fetchUserBorrowedBooks } = mapDispatchToProps(dispatch);
    fetchUserBorrowedBooks();
    expect(dispatch).toHaveBeenCalled();
  });

  test('mapStateToProps', () => {
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('borrowedBooks');
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('userBorrowedBookCount');
  });

})