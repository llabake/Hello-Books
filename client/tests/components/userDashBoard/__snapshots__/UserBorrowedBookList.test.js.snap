// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The User Borrowed Book List table should render and not crash 1`] = `
<div>
  <NothingFound
    text="Start borrowing books today!!!"
  />
</div>
`;
